mountpoint-s3-client/src/s3_crt_client.rs

```diff
@@ -630,9 +630,46 @@ fn extract_range_header(headers: &Headers) -> Option<Range<u64>> {
     Some(start..end + 1)
 }
 
+#[derive(Debug)]
+pub struct S3PutObjectRequest {
+    body: S3HttpRequest<Vec<u8>, PutObjectError>,
+    writer: AsyncStreamWriter,
+}
+
+impl S3PutObjectRequest {
+    fn new(body: S3HttpRequest<Vec<u8>, PutObjectError>, writer: AsyncStreamWriter) -> Self {
+        Self { body, writer }
+    }
+}
+
+#[async_trait]
+impl PutObjectRequest for S3PutObjectRequest {
+    type ClientError = S3RequestError;
+
+    async fn write(&mut self, slice: &[u8]) -> ObjectClientResult<(), PutObjectError, Self::ClientError> {
+        self.writer
+            .write(slice)
+            .await
+            .map_err(|e| S3RequestError::InternalError(Box::new(e)).into())
+    }
+
+    async fn complete(mut self) -> ObjectClientResult<PutObjectResult, PutObjectError, Self::ClientError> {
```

Comment:
Should we add rustdoc on methods like complete here because it can be difficult what AsyncStreamWriter is doing (might be easier with VSCode though)?